### PR TITLE
Bugfix objective function initialization

### PR DESCRIPTION
**Pull request by bvanessen**
Fixing a bug where the objective function is not getting setup in the model setup.  Also adding objective function state into the checkpoint.

### CODE REVIEW

**Comment by bvanessen** [comment]
@timmoon10 @benson31 This is ready for re-review.

**Comment by bvanessen** [comment]
@timmoon10 @benson31 Can you guys checkout the new approximate diff function that I added to this PR to allow us to ignore small differences between two floats in an XML checkpoint function.

**Comment by bvanessen** [comment]
> I don't see anything breaking in the approximate diff function. Maybe we could replace the hard-coded tolerance of `1e-7` with a function argument, which could be helpful if we run a test with fp16.

I added an argument to pass the threshold into the function and print the threshold used.

File: include/lbann/layers/transform/evaluation.hpp [hunk 1]
```diff
@@ -35,7 +35,12 @@ namespace lbann {
 template <typename TensorDataType>
 class abstract_evaluation_layer : public transform_layer<TensorDataType> {
 public:
-  using CPUMatType = El::Matrix<TensorDataType, El::Device::CPU>;
+#ifdef LBANN_DETERMINISTIC
+  using EvalDataType = EvalType;
```
[the remark this answers is not in the document]
Comment (timmoon10) [reply include/lbann/layers/transform/evaluation.hpp:39]: I suppose using global typedefs is problematic. It now makes sense to me for the objective function to be templated on `EvalType` and the evaluation layer would be templated both on `DataType` and `EvalType`.

**Pull request merged**